### PR TITLE
Rearrange page search

### PR DESCRIPTION
**Pull request by sascha-karnatz**
Previously the page and the ingredients (and in previous versions also essences) were stored as separate PgSearch::Document entries. On search these document were combined to a single document to make it usable in the search itself. These mechanic made it pretty difficult to extend the search with other models. In newer versions of Alchemy the page is only available after a new page version is released and these mechanic is used to create a single PgSearch::Document with the content of the page and all supported ingredients. This change makes the whole search and the index creation less complex.

### CODE REVIEW

File: app/extensions/alchemy/pg_search/ingredient_extension.rb [hunk 1]
```diff
@@ -1,18 +1,15 @@
 module Alchemy::PgSearch::IngredientExtension
+  def searchable_content
+    content_field = case type
+      when "Alchemy::Ingredients::Richtext"
```
Comment (tvdeyen) [inline app/extensions/alchemy/pg_search/ingredient_extension.rb]: I think it makes sense to make this a Map as a replacement for `SEARCHABLE_INGREDIENTS`

```rb
Alchemy.searchable_ingredients = {
  "Alchemy::Ingredients::Richtext" => :stripped_body,
  "Alchemy::Ingredients::Picture" => :caption
}
```

File: app/extensions/alchemy/pg_search/ingredient_extension.rb [hunk 2]
```diff
@@ -1,18 +1,15 @@
...
-    }
-  end
-
-  def self.prepended(base)
-    base.include PgSearch::Model
-    base.multisearchable(multisearch_config)
+    send(content_field)
   end
 
   def searchable?
```
Comment (tvdeyen) [inline app/extensions/alchemy/pg_search/ingredient_extension.rb]: We can remove `Alchemy::PgSearch.is_searchable?(type)`

File: app/extensions/alchemy/pg_search/ingredient_extension.rb [hunk 3]
```diff
@@ -1,18 +1,15 @@
 module Alchemy::PgSearch::IngredientExtension
```
Comment (tvdeyen) [inline app/extensions/alchemy/pg_search/ingredient_extension.rb]: This module could be a more generic `Alchemy::Search::Searchable` module

File: app/extensions/alchemy/pg_search/ingredient_extension.rb [hunk 4]
```diff
@@ -24,7 +21,3 @@ def searchable?
 
 # add the PgSearch model to all ingredients
 Alchemy::Ingredient.prepend(Alchemy::PgSearch::IngredientExtension)
```
Comment (tvdeyen) [inline app/extensions/alchemy/pg_search/ingredient_extension.rb]:
```rb
Alchemy::Ingredient.prepend(Alchemy::Search::Searchable)
```

File: lib/alchemy-pg_search.rb [hunk 5]
```diff
@@ -19,19 +19,17 @@ def self.is_searchable?(ingredient_type)
     end
 
     ##
-    # index all supported Alchemy models
+    # index all supported Alchemy pages
     def self.rebuild
-      [Alchemy::Page, Alchemy::Ingredient].each do |model|
-        ::PgSearch::Multisearch.rebuild(model)
-      end
+      Alchemy::Page.all.each{ |page| index_page(page) }
```
Comment (tvdeyen) [inline lib/alchemy-pg_search.rb]: we should wrap this in a transaction

File: lib/alchemy-pg_search.rb [hunk 6]
```diff
@@ -41,12 +39,16 @@ def self.remove_page(page)
...
+      return if document.nil?
+      
       page.all_elements.includes(:ingredients).find_each do |element|
-        element.ingredients.select { |i| Alchemy::PgSearch.is_searchable?(i.type) }.each do |ingredient|
-          ingredient.update_pg_search_document
+        element.ingredients.select { |i| i.searchable? }.each do |ingredient|
+          document.content += " #{ingredient.searchable_content}"
         end
       end
+      document.content.strip!
```
Comment (tvdeyen) [inline lib/alchemy-pg_search.rb]:
```rb
document.content.squish!
```

File: app/extensions/alchemy/pg_search/ingredient_extension.rb [hunk 7]
```diff
@@ -1,18 +1,15 @@
 module Alchemy::PgSearch::IngredientExtension
+  def searchable_content
+    content_field = case type
+      when "Alchemy::Ingredients::Richtext"
+        :stripped_body
+      when "Alchemy::Ingredients::Picture"
+        :caption
+      else
+        :value
+      end
```
Comment (tvdeyen) [inline app/extensions/alchemy/pg_search/ingredient_extension.rb]: We should add `squish` here

File: lib/alchemy-pg_search.rb [hunk 8]
```diff
@@ -41,12 +39,16 @@ def self.remove_page(page)
...
       remove_page page
 
-      page.update_pg_search_document
+      document = page.update_pg_search_document
+      return if document.nil?
+      
       page.all_elements.includes(:ingredients).find_each do |element|
-        element.ingredients.select { |i| Alchemy::PgSearch.is_searchable?(i.type) }.each do |ingredient|
-          ingredient.update_pg_search_document
+        element.ingredients.select { |i| i.searchable? }.each do |ingredient|
```
Comment (tvdeyen) [inline lib/alchemy-pg_search.rb]: We could use `map { _1.searchable_content }.join(" ")` here.

Comment (tvdeyen) [reply lib/alchemy-pg_search.rb]: and maybe add `ingredients.searchable` scope

Comment (tvdeyen) [reply lib/alchemy-pg_search.rb]: and add a `elements.searchable` scope as well.

File: lib/alchemy-pg_search.rb [hunk 9]
```diff
@@ -6,7 +6,7 @@ module Alchemy
   @@search_class = PgSearch
 
   module PgSearch
-    SEARCHABLE_INGREDIENTS = %w[Text Richtext Picture]
+    SEARCHABLE_INGREDIENTS = %w[Text Richtext Headline Picture]
```
Comment (tvdeyen) [inline lib/alchemy-pg_search.rb]: 👍🏻

File: lib/alchemy-pg_search.rb [hunk 10]
```diff
@@ -41,12 +39,15 @@ def self.remove_page(page)
     def self.index_page(page)
       remove_page page
 
-      page.update_pg_search_document
+      document = page.update_pg_search_document
+      return if document.nil?
+      
       page.all_elements.includes(:ingredients).find_each do |element|
```
Comment (tvdeyen) [inline lib/alchemy-pg_search.rb]: Maybe we have an N+1 here

```rb
page.all_elements.includes(ingredients: {element: {page_version: :page}}).
```

File: lib/alchemy-pg_search.rb [hunk 11]
```diff
@@ -57,12 +58,16 @@ def self.index_page(page)
...
-                        .joins(:page)
 
-      query = query.merge(Alchemy::Page.accessible_by(ability, :read)) if ability
+      if ability
+        # left_joins method is not usable here, because the order of the joins are incorrect
+        # and would result in a SQL error. We can receive the correct query order with these
+        # odd left join string
+        # Ref: https://guides.rubyonrails.org/active_record_querying.html#using-a-string-sql-fragment
+        query = query
+                  .joins("LEFT JOIN alchemy_pages ON alchemy_pages.id = pg_search_documents.page_id")
```
Comment (tvdeyen) [inline lib/alchemy-pg_search.rb:68]: Do we need the `left_join` outside of the `if` statement?

File: lib/alchemy-pg_search.rb (hunk 10 above)
Comment (sascha-karnatz) [reply lib/alchemy-pg_search.rb]: changed it and it is working without n+1s

File: lib/alchemy-pg_search.rb (hunk 11 above)
Comment (sascha-karnatz) [reply lib/alchemy-pg_search.rb:68]: It would make the query slower and Rails would reload the page again.

Comment (tvdeyen) [reply lib/alchemy-pg_search.rb:68]: So, you are saying with a `ability` passed in we get N+1s?

File: lib/alchemy-pg_search.rb [hunk 12]
```diff
@@ -5,24 +5,23 @@ module Alchemy
...
-    end
-
-    ##
-    # index all supported Alchemy models
+    # index all supported Alchemy pages
     def self.rebuild
-      [Alchemy::Page, Alchemy::Ingredient].each do |model|
-        ::PgSearch::Multisearch.rebuild(model)
+      ActiveRecord::Base.transaction do
+        ::PgSearch::Document.delete_by(searchable_type: "Alchemy::Page")
```
Comment (tvdeyen) [inline lib/alchemy-pg_search.rb]:
```suggestion
        ::PgSearch::Document.delete_all
```

File: lib/alchemy-pg_search.rb [hunk 13]
```diff
@@ -41,12 +40,15 @@ def self.remove_page(page)
     def self.index_page(page)
       remove_page page
```
Comment (tvdeyen) [inline lib/alchemy-pg_search.rb]: We do not need this anymore

```suggestion
```

File: lib/alchemy-pg_search.rb [hunk 14]
```diff
@@ -39,14 +38,16 @@ def self.remove_page(page)
...
-      end
+
+      document = page.pg_search_document
+      return if document.nil?
+
+      ingredient_content = page.all_elements.includes(ingredients: {element: :page}).map do |element|
+        element.ingredients.select { |i| i.searchable? }.map(&:searchable_content).join(" ")
+      end.join(" ")
+
+      document.update(content: "#{document.content} #{ingredient_content}".squish)
```
Comment (tvdeyen) [inline lib/alchemy-pg_search.rb]: Maybe we need to `update_columns`

```suggestion
      document.update_columns(content: "#{document.content} #{ingredient_content}".squish)
```

**Review by tvdeyen — APPROVED** [review]

**Pull request merged**